contracts/src/tokens/factory.cairo

```diff
@@ -0,0 +1,167 @@
+
+    // Components.
+    component!(path: OwnableComponent, storage: ownable, event: OwnableEvent);
+    impl OwnableInternalImpl = OwnableComponent::InternalImpl<ContractState>;
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        MemeCoinCreated: MemeCoinCreated,
+        #[flat]
+        OwnableEvent: OwnableComponent::Event
+    }
+
+    #[derive(Drop, starknet::Event)]
+    struct MemeCoinCreated {
+        owner: ContractAddress,
+        initial_recipient: ContractAddress,
+        name: felt252,
+        symbol: felt252,
+        initial_supply: u256,
+        memecoin_address: ContractAddress
+    }
+
+    #[storage]
+    struct Storage {
+        // Components.
+        #[substorage(v0)]
+        ownable: OwnableComponent::Storage,
+        memecoin_class_hash: ClassHash,
+        amms: LegacyMap<u32, AMM>,
+        amms_len: u32
+    }
+
+    #[constructor]
+    fn constructor(
+        ref self: ContractState,
+        owner: ContractAddress,
+        memecoin_class_hash: ClassHash,
+        amms: Array<AMM>
+    ) {
+        // Initialize the owner.
+        self.ownable.initializer(owner);
+        self.memecoin_class_hash.write(memecoin_class_hash);
+
+        let mut i = 0;
+        let amms_len = amms.len();
+        loop {
+            if amms_len == i {
+                break;
+            }
+            let amm = *amms[i];
+            self.amms.write(i, amm);
+            i += 1;
+        };
+        self.amms_len.write(amms_len);
+    }
+
+    #[external(v0)]
+    impl UnruggableMemeCoinFactoryImpl of IUnruggableMemecoinFactory<ContractState> {
+        fn create_memecoin(
+            ref self: ContractState,
+            owner: ContractAddress,
+            initial_recipient: ContractAddress,
+            name: felt252,
+            symbol: felt252,
+            initial_supply: u256
+        ) -> ContractAddress {
+            let contract_address_salt = generate_salt(owner, name, symbol);
+
+            // General calldata
+            let mut calldata = serialize_calldata(
+                owner, initial_recipient, name, symbol, initial_supply
+            );
+
+            Serde::serialize(@self.whitelisted_amms().into(), ref calldata);
+
+            let (memecoin_address, _) = deploy_syscall(
+                self.memecoin_class_hash.read(), contract_address_salt, calldata.span(), false
+            )
+                .unwrap_syscall();
+
+            self
+                .emit(
+                    MemeCoinCreated {
+                        owner, name, symbol, initial_recipient, initial_supply, memecoin_address
+                    }
+                );
+            memecoin_address
+        }
+    }
+
+    #[generate_trait]
+    impl InternalFunctions of InternalFunctionsTrait {
+        fn whitelisted_amms(self: @ContractState) -> Array<AMM> {
+            let mut amms = array![];
+            let amms_len = self.amms_len.read();
+            let mut i = 0;
+
+            loop {
```

Comment:
`match arr.pop_front()` based loop

Reply:
How could I apply it to this case? I'm reading the storage, which is a legacy map.
`amm_configs: LegacyMap<felt252, ContractAddress>`